extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/LazyObjectSummariesIterator.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.storage.s3;
+
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import org.apache.druid.java.util.common.RE;
+
+import java.net.URI;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * Iterator class used by {@link S3Utils#lazyObjectSummaryIterator}.
+ *
+ * Walks a set of prefixes, returning all objects underneath them except for directory placeholders.
+ */
+public class LazyObjectSummariesIterator implements Iterator<S3ObjectSummary>
+{
+  private final ServerSideEncryptingAmazonS3 s3Client;
+  private final Iterator<URI> prefixesIterator;
+  private final int maxListingLength;
+
+  private ListObjectsV2Request request;
+  private ListObjectsV2Result result;
+  private Iterator<S3ObjectSummary> objectSummaryIterator;
+  private S3ObjectSummary currentObjectSummary;
+
+  LazyObjectSummariesIterator(
+      final ServerSideEncryptingAmazonS3 s3Client,
+      final Iterable<URI> prefixes,
+      final int maxListingLength
+  )
+  {
+    this.s3Client = s3Client;
+    this.prefixesIterator = prefixes.iterator();
+    this.maxListingLength = maxListingLength;
+
+    prepareNextRequest();
+    fetchNextBatch();
+    advanceObjectSummary();
+  }
+
+  private void prepareNextRequest()
+  {
+    final URI currentUri = prefixesIterator.next();
+    final String currentBucket = currentUri.getAuthority();
+    final String currentPrefix = S3Utils.extractS3Key(currentUri);
+
+    request = new ListObjectsV2Request()
+        .withBucketName(currentBucket)
+        .withPrefix(currentPrefix)
+        .withMaxKeys(maxListingLength);
+  }
+
+  private void fetchNextBatch()
+  {
+    try {
+      result = S3Utils.retryS3Operation(() -> s3Client.listObjectsV2(request));
+      request.setContinuationToken(result.getNextContinuationToken());
+      objectSummaryIterator = result.getObjectSummaries().iterator();
+    }
+    catch (Exception e) {
+      throw new RE(
+          e,
+          "Failed to get object summaries from S3 bucket[%s], prefix[%s]",
```

Comment:
This log message doesn't tell us if the error is because of permissions or something else. Would we want to expose that information in here?

Reply:
It would appear if this thrown exception is fully logged, since its cause will be set to `e`, which has a message that includes the stack trace. But I agree it would be nicer to include the S3 error in _this_ exception's message rather than only as its cause. So I added it in the latest patch.